### PR TITLE
etcdmain: log multiple listener errors

### LINKED ISSUE #10652: etcdmain: accept tcp [::]:2379: use of closed network connection when using --cipher-suites without TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
I'm trying to enable TLS on etcd 3.3.12. In particular, I'm trying to use the `--cipher-suites` option added in v3.3.7 (https://github.com/etcd-io/etcd/blob/f29b1ada19713544b698dab8c94c97cfa1e83dac/CHANGELOG-3.3.md#security-authentication-2). However, I've found that if I specify some limited cipher suite, etcd immediately exits with `etcdmain: accept tcp [::]:2379: use of closed network connection
etcd`.

- "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256" by itself is fine
- "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384" by itself is not fine
- "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256" is fine
- "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305" is not fine

I have a script that repros this using the etcd Docker image:

```
$ cat etcd_tls_repro.sh
export HostIP="192.168.56.101"

run_etcd() {
  ciphers=$1
  docker run -v /usr/share/ca-certificates/:/etc/ssl/certs --net=host \
   --name etcd quay.io/coreos/etcd:v3.3.12 /usr/local/bin/etcd \
   -name etcd0 \
   -advertise-client-urls https://${HostIP}:2379,https://${HostIP}:4001 \
   -listen-client-urls https://0.0.0.0:2379,https://0.0.0.0:4001 \
   -initial-advertise-peer-urls https://${HostIP}:2380 \
   -listen-peer-urls https://0.0.0.0:2380 \
   -initial-cluster-token etcd-cluster-1 \
   -initial-cluster etcd0=https://${HostIP}:2380 \
   -initial-cluster-state new \
   --auto-tls --peer-auto-tls \
   --debug \
   --cipher-suites="$ciphers"
  docker rm etcd
}

echo "***"
echo "Run with 1 cipher"
echo "***"
run_etcd "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
echo "***"
echo "Run with 2 ciphers"
echo "***"
run_etcd "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"



$ bash etcd_tls_repro.sh
***
Run with 1 cipher
***
2019-04-16 13:45:36.395714 I | etcdmain: etcd Version: 3.3.12
2019-04-16 13:45:36.395761 I | etcdmain: Git SHA: d57e8b8
2019-04-16 13:45:36.395765 I | etcdmain: Go Version: go1.10.8
2019-04-16 13:45:36.395768 I | etcdmain: Go OS/Arch: linux/amd64
2019-04-16 13:45:36.395771 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
2019-04-16 13:45:36.395777 W | etcdmain: no data-dir provided, using default data-dir ./etcd0.etcd
2019-04-16 13:45:36.429820 I | embed: peerTLS: cert = etcd0.etcd/fixtures/peer/cert.pem, key = etcd0.etcd/fixtures/peer/key.pem, ca = , trusted-ca = , client-cert-auth = false, crl-file =
2019-04-16 13:45:36.444011 I | embed: listening for peers on https://0.0.0.0:2380
2019-04-16 13:45:36.471300 I | embed: listening for client requests on 0.0.0.0:2379
2019-04-16 13:45:36.471535 I | embed: listening for client requests on 0.0.0.0:4001
2019-04-16 13:45:36.498658 I | etcdserver: name = etcd0
2019-04-16 13:45:36.498680 I | etcdserver: data dir = etcd0.etcd
2019-04-16 13:45:36.498686 I | etcdserver: member dir = etcd0.etcd/member
2019-04-16 13:45:36.498690 I | etcdserver: heartbeat = 100ms
2019-04-16 13:45:36.498716 I | etcdserver: election = 1000ms
2019-04-16 13:45:36.498722 I | etcdserver: snapshot count = 100000
2019-04-16 13:45:36.498736 I | etcdserver: advertise client URLs = https://192.168.56.101:2379,https://192.168.56.101:4001
2019-04-16 13:45:36.498742 I | etcdserver: initial advertise peer URLs = https://192.168.56.101:2380
2019-04-16 13:45:36.498788 I | etcdserver: initial cluster = etcd0=https://192.168.56.101:2380
2019-04-16 13:45:36.517406 I | etcdserver: starting member 30d21c9e60dd6e27 in cluster 6f5958f2a13709b1
2019-04-16 13:45:36.517438 I | raft: 30d21c9e60dd6e27 became follower at term 0
2019-04-16 13:45:36.517446 I | raft: newRaft 30d21c9e60dd6e27 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2019-04-16 13:45:36.517450 I | raft: 30d21c9e60dd6e27 became follower at term 1
2019-04-16 13:45:36.531041 W | auth: simple token is not cryptographically signed
2019-04-16 13:45:36.567743 I | etcdserver: starting server... [version: 3.3.12, cluster version: to_be_decided]
2019-04-16 13:45:36.569764 I | etcdserver: 30d21c9e60dd6e27 as single-node; fast-forwarding 9 ticks (election ticks 10)
2019-04-16 13:45:36.570880 I | etcdserver/membership: added member 30d21c9e60dd6e27 [https://192.168.56.101:2380] to cluster 6f5958f2a13709b1
2019-04-16 13:45:36.585215 I | embed: ClientTLS: cert = etcd0.etcd/fixtures/client/cert.pem, key = etcd0.etcd/fixtures/client/key.pem, ca = , trusted-ca = , client-cert-auth = false, crl-file =
2019-04-16 13:45:37.318647 I | raft: 30d21c9e60dd6e27 is starting a new election at term 1
2019-04-16 13:45:37.318677 I | raft: 30d21c9e60dd6e27 became candidate at term 2
2019-04-16 13:45:37.318701 I | raft: 30d21c9e60dd6e27 received MsgVoteResp from 30d21c9e60dd6e27 at term 2
2019-04-16 13:45:37.318954 I | raft: 30d21c9e60dd6e27 became leader at term 2
2019-04-16 13:45:37.318968 I | raft: raft.node: 30d21c9e60dd6e27 elected leader 30d21c9e60dd6e27 at term 2
2019-04-16 13:45:37.319485 I | etcdserver: setting up the initial cluster version to 3.3
2019-04-16 13:45:37.319686 I | etcdserver: published {Name:etcd0 ClientURLs:[https://192.168.56.101:2379 https://192.168.56.101:4001]} to cluster 6f5958f2a13709b1
2019-04-16 13:45:37.319903 I | embed: ready to serve client requests
2019-04-16 13:45:37.323766 I | embed: ready to serve client requests
2019-04-16 13:45:37.339387 N | etcdserver/membership: set the initial cluster version to 3.3
2019-04-16 13:45:37.350272 I | etcdserver/api: enabled capabilities for version 3.3
INFO: 2019/04/16 13:45:37 dialing to target with scheme: ""
INFO: 2019/04/16 13:45:37 could not get resolver for scheme: ""
INFO: 2019/04/16 13:45:37 dialing to target with scheme: ""
INFO: 2019/04/16 13:45:37 could not get resolver for scheme: ""
2019-04-16 13:45:37.376420 I | embed: serving client requests on [::]:2379
2019-04-16 13:45:37.383581 C | etcdmain: accept tcp [::]:2379: use of closed network connection
etcd
***
Run with 2 ciphers
***
2019-04-16 13:45:38.165088 I | etcdmain: etcd Version: 3.3.12
2019-04-16 13:45:38.165162 I | etcdmain: Git SHA: d57e8b8
2019-04-16 13:45:38.165166 I | etcdmain: Go Version: go1.10.8
2019-04-16 13:45:38.165169 I | etcdmain: Go OS/Arch: linux/amd64
2019-04-16 13:45:38.165173 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
2019-04-16 13:45:38.165179 W | etcdmain: no data-dir provided, using default data-dir ./etcd0.etcd
2019-04-16 13:45:38.206865 I | embed: peerTLS: cert = etcd0.etcd/fixtures/peer/cert.pem, key = etcd0.etcd/fixtures/peer/key.pem, ca = , trusted-ca = , client-cert-auth = false, crl-file =
2019-04-16 13:45:38.218970 I | embed: listening for peers on https://0.0.0.0:2380
2019-04-16 13:45:38.243689 I | embed: listening for client requests on 0.0.0.0:2379
2019-04-16 13:45:38.243758 I | embed: listening for client requests on 0.0.0.0:4001
2019-04-16 13:45:38.270098 I | etcdserver: name = etcd0
2019-04-16 13:45:38.270189 I | etcdserver: data dir = etcd0.etcd
2019-04-16 13:45:38.270197 I | etcdserver: member dir = etcd0.etcd/member
2019-04-16 13:45:38.270201 I | etcdserver: heartbeat = 100ms
2019-04-16 13:45:38.270204 I | etcdserver: election = 1000ms
2019-04-16 13:45:38.270208 I | etcdserver: snapshot count = 100000
2019-04-16 13:45:38.270217 I | etcdserver: advertise client URLs = https://192.168.56.101:2379,https://192.168.56.101:4001
2019-04-16 13:45:38.270221 I | etcdserver: initial advertise peer URLs = https://192.168.56.101:2380
2019-04-16 13:45:38.270230 I | etcdserver: initial cluster = etcd0=https://192.168.56.101:2380
2019-04-16 13:45:38.293613 I | etcdserver: starting member 30d21c9e60dd6e27 in cluster 6f5958f2a13709b1
2019-04-16 13:45:38.293647 I | raft: 30d21c9e60dd6e27 became follower at term 0
2019-04-16 13:45:38.293656 I | raft: newRaft 30d21c9e60dd6e27 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2019-04-16 13:45:38.293660 I | raft: 30d21c9e60dd6e27 became follower at term 1
2019-04-16 13:45:38.304529 W | auth: simple token is not cryptographically signed
2019-04-16 13:45:38.337386 I | etcdserver: starting server... [version: 3.3.12, cluster version: to_be_decided]
2019-04-16 13:45:38.338245 I | etcdserver: 30d21c9e60dd6e27 as single-node; fast-forwarding 9 ticks (election ticks 10)
2019-04-16 13:45:38.338761 I | etcdserver/membership: added member 30d21c9e60dd6e27 [https://192.168.56.101:2380] to cluster 6f5958f2a13709b1
2019-04-16 13:45:38.355113 I | embed: ClientTLS: cert = etcd0.etcd/fixtures/client/cert.pem, key = etcd0.etcd/fixtures/client/key.pem, ca = , trusted-ca = , client-cert-auth = false, crl-file =
2019-04-16 13:45:38.794384 I | raft: 30d21c9e60dd6e27 is starting a new election at term 1
2019-04-16 13:45:38.794645 I | raft: 30d21c9e60dd6e27 became candidate at term 2
2019-04-16 13:45:38.794719 I | raft: 30d21c9e60dd6e27 received MsgVoteResp from 30d21c9e60dd6e27 at term 2
2019-04-16 13:45:38.794818 I | raft: 30d21c9e60dd6e27 became leader at term 2
2019-04-16 13:45:38.794919 I | raft: raft.node: 30d21c9e60dd6e27 elected leader 30d21c9e60dd6e27 at term 2
2019-04-16 13:45:38.795298 I | etcdserver: setting up the initial cluster version to 3.3
2019-04-16 13:45:38.795580 I | etcdserver: published {Name:etcd0 ClientURLs:[https://192.168.56.101:2379 https://192.168.56.101:4001]} to cluster 6f5958f2a13709b1
2019-04-16 13:45:38.795659 I | embed: ready to serve client requests
2019-04-16 13:45:38.797356 I | embed: ready to serve client requests
2019-04-16 13:45:38.807807 N | etcdserver/membership: set the initial cluster version to 3.3
2019-04-16 13:45:38.808164 I | etcdserver/api: enabled capabilities for version 3.3
INFO: 2019/04/16 13:45:38 dialing to target with scheme: ""
INFO: 2019/04/16 13:45:38 could not get resolver for scheme: ""
INFO: 2019/04/16 13:45:38 dialing to target with scheme: ""
INFO: 2019/04/16 13:45:38 could not get resolver for scheme: ""
2019-04-16 13:45:38.843507 I | embed: serving client requests on [::]:4001
2019-04-16 13:45:38.844635 I | embed: serving client requests on [::]:2379
WARNING: 2019/04/16 13:45:38 Failed to dial 0.0.0.0:2379: connection error: desc = "transport: authentication handshake failed: remote error: tls: handshake failure"; please retry.
WARNING: 2019/04/16 13:45:38 Failed to dial 0.0.0.0:2379: connection error: desc = "transport: authentication handshake failed: remote error: tls: handshake failure"; please retry.
WARNING: 2019/04/16 13:45:38 Failed to dial 0.0.0.0:4001: connection error: desc = "transport: authentication handshake failed: remote error: tls: handshake failure"; please retry.
WARNING: 2019/04/16 13:45:38 Failed to dial 0.0.0.0:4001: connection error: desc = "transport: authentication handshake failed: remote error: tls: handshake failure"; please retry.
^C2019-04-16 13:45:51.460770 N | pkg/osutil: received interrupt signal, shutting down...
2019-04-16 13:45:51.461059 I | etcdserver: skipped leadership transfer for single member cluster
```

### PR DESCRIPTION
**Pull request by rkday**
When a set of cipher suites are provided that aren't accepted by the
http2 library, the error log indicating this is suppressed, because
another listener fails first (with a connection error) and we exit when
the first listener fails.

Instead, we now collect and log all listener error messages before
exiting.

Fixes #10652 .

There might be a more idiomatic way to do this - I confess I'm not very familiar with Go - but this change means that rather than:

```
$ ~/go/bin/etcd -advertise-client-urls https://127.0.0.1:2379,https://127.0.0.1:4001 -listen-client-urls https://0.0.0.0:2379,https://0.0.0.0:4001 -initial-advertise-peer-urls https://127.0.0.1:2380 -listen-peer-urls https://0.0.0.0:2380 -initial-cluster-token etcd-cluster-1 -initial-cluster etcd0=https://127.0.0.1:2380 -initial-cluster-state new --auto-tls --peer-auto-tls --cipher-suites=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
2019-04-28 20:37:58.993302 I | etcdmain: etcd Version: 3.3.0+git
2019-04-28 20:37:58.993430 I | etcdmain: Git SHA: Not provided (use ./build instead of go build)
2019-04-28 20:37:58.993449 I | etcdmain: Go Version: go1.10.4
2019-04-28 20:37:58.993466 I | etcdmain: Go OS/Arch: linux/amd64
2019-04-28 20:37:58.993483 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2019-04-28 20:37:58.993517 W | etcdmain: no data-dir provided, using default data-dir ./default.etcd
2019-04-28 20:37:58.993650 W | etcdmain: found invalid file/dir fixtures under data dir default.etcd (Ignore this if you are upgrading etcd)
2019-04-28 20:37:58.993676 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2019-04-28 20:37:58.993791 I | embed: peerTLS: cert = default.etcd/fixtures/peer/cert.pem, key = default.etcd/fixtures/peer/key.pem, trusted-ca = , client-cert-auth = false, crl-file = 
2019-04-28 20:37:59.013160 I | embed: name = default
2019-04-28 20:37:59.013182 I | embed: data dir = default.etcd
2019-04-28 20:37:59.013199 I | embed: member dir = default.etcd/member
2019-04-28 20:37:59.013207 I | embed: heartbeat = 100ms
2019-04-28 20:37:59.013216 I | embed: election = 1000ms
2019-04-28 20:37:59.013226 I | embed: snapshot count = 100000
2019-04-28 20:37:59.013251 I | embed: advertise client URLs = https://127.0.0.1:2379,https://127.0.0.1:4001
2019-04-28 20:37:59.013261 I | embed: initial advertise peer URLs = https://127.0.0.1:2380
2019-04-28 20:37:59.013271 I | embed: initial cluster = 
2019-04-28 20:37:59.027693 I | etcdserver: restarting member 8e9e05c52164694d in cluster cdf818194e3a8c32 at commit index 66
{"level":"info","ts":"2019-04-28T20:37:59.027+0100","caller":"raft/raft.go:712","msg":"8e9e05c52164694d became follower at term 33"}
{"level":"info","ts":"2019-04-28T20:37:59.027+0100","caller":"raft/raft.go:389","msg":"newRaft 8e9e05c52164694d [peers: [], term: 33, commit: 66, applied: 0, lastindex: 66, lastterm: 33]"}
2019-04-28 20:37:59.040275 W | auth: simple token is not cryptographically signed
2019-04-28 20:37:59.076055 I | etcdserver: starting server... [version: 3.3.0+git, cluster version: to_be_decided]
2019-04-28 20:37:59.076767 I | etcdserver/membership: added member 8e9e05c52164694d [http://localhost:2380] to cluster cdf818194e3a8c32
2019-04-28 20:37:59.076900 N | etcdserver/membership: set the initial cluster version to 3.3
2019-04-28 20:37:59.076951 I | etcdserver/api: enabled capabilities for version 3.3
2019-04-28 20:37:59.092579 I | embed: ClientTLS: cert = default.etcd/fixtures/client/cert.pem, key = default.etcd/fixtures/client/key.pem, trusted-ca = , client-cert-auth = false, crl-file = 
2019-04-28 20:37:59.092659 I | embed: listening for peers on [::]:2380
{"level":"info","ts":"2019-04-28T20:38:00.428+0100","caller":"raft/raft.go:922","msg":"8e9e05c52164694d is starting a new election at term 33"}
{"level":"info","ts":"2019-04-28T20:38:00.428+0100","caller":"raft/raft.go:725","msg":"8e9e05c52164694d became candidate at term 34"}
{"level":"info","ts":"2019-04-28T20:38:00.428+0100","caller":"raft/raft.go:820","msg":"8e9e05c52164694d received MsgVoteResp from 8e9e05c52164694d at term 34"}
{"level":"info","ts":"2019-04-28T20:38:00.429+0100","caller":"raft/raft.go:777","msg":"8e9e05c52164694d became leader at term 34"}
{"level":"info","ts":"2019-04-28T20:38:00.429+0100","caller":"raft/node.go:330","msg":"raft.node: 8e9e05c52164694d elected leader 8e9e05c52164694d at term 34"}
2019-04-28 20:38:00.435967 I | embed: ready to serve client requests
2019-04-28 20:38:00.436289 I | etcdserver: published {Name:default ClientURLs:[https://127.0.0.1:2379 https://127.0.0.1:4001]} to cluster cdf818194e3a8c32
2019-04-28 20:38:00.436429 I | embed: ready to serve client requests
2019-04-28 20:38:00.473056 I | embed: serving client requests on [::]:4001
2019-04-28 20:38:00.473273 C | etcdmain: accept tcp [::]:4001: use of closed network connection
```

I get:

```
$ ./etcd -advertise-client-urls https://127.0.0.1:2379,https://127.0.0.1:4001 -listen-client-urls https://0.0.0.0:2379,https://0.0.0.0:4001 -initial-advertise-peer-urls https://127.0.0.1:2380 -listen-peer-urls https://0.0.0.0:2380 -initial-cluster-token etcd-cluster-1 -initial-cluster etcd0=https://127.0.0.1:2380 -initial-cluster-state new --auto-tls --peer-auto-tls --cipher-suites=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
2019-04-28 20:39:45.420471 I | etcdmain: etcd Version: 3.3.0+git
2019-04-28 20:39:45.420653 I | etcdmain: Git SHA: Not provided (use ./build instead of go build)
2019-04-28 20:39:45.420683 I | etcdmain: Go Version: go1.10.4
2019-04-28 20:39:45.420690 I | etcdmain: Go OS/Arch: linux/amd64
2019-04-28 20:39:45.420713 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2019-04-28 20:39:45.420773 W | etcdmain: no data-dir provided, using default data-dir ./default.etcd
2019-04-28 20:39:45.420879 W | etcdmain: found invalid file/dir fixtures under data dir default.etcd (Ignore this if you are upgrading etcd)
2019-04-28 20:39:45.420905 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2019-04-28 20:39:45.421008 I | embed: peerTLS: cert = default.etcd/fixtures/peer/cert.pem, key = default.etcd/fixtures/peer/key.pem, trusted-ca = , client-cert-auth = false, crl-file = 
2019-04-28 20:39:45.438734 I | embed: name = default
2019-04-28 20:39:45.438760 I | embed: data dir = default.etcd
2019-04-28 20:39:45.438768 I | embed: member dir = default.etcd/member
2019-04-28 20:39:45.438775 I | embed: heartbeat = 100ms
2019-04-28 20:39:45.438781 I | embed: election = 1000ms
2019-04-28 20:39:45.438789 I | embed: snapshot count = 100000
2019-04-28 20:39:45.438814 I | embed: advertise client URLs = https://127.0.0.1:2379,https://127.0.0.1:4001
2019-04-28 20:39:45.438825 I | embed: initial advertise peer URLs = https://127.0.0.1:2380
2019-04-28 20:39:45.438836 I | embed: initial cluster = 
2019-04-28 20:39:45.457228 I | etcdserver: restarting member 8e9e05c52164694d in cluster cdf818194e3a8c32 at commit index 70
{"level":"info","ts":"2019-04-28T20:39:45.457+0100","caller":"raft/raft.go:712","msg":"8e9e05c52164694d became follower at term 35"}
{"level":"info","ts":"2019-04-28T20:39:45.457+0100","caller":"raft/raft.go:389","msg":"newRaft 8e9e05c52164694d [peers: [], term: 35, commit: 70, applied: 0, lastindex: 70, lastterm: 35]"}
2019-04-28 20:39:45.476368 W | auth: simple token is not cryptographically signed
2019-04-28 20:39:45.522689 I | etcdserver: starting server... [version: 3.3.0+git, cluster version: to_be_decided]
2019-04-28 20:39:45.523659 I | etcdserver/membership: added member 8e9e05c52164694d [http://localhost:2380] to cluster cdf818194e3a8c32
2019-04-28 20:39:45.523805 N | etcdserver/membership: set the initial cluster version to 3.3
2019-04-28 20:39:45.523880 I | etcdserver/api: enabled capabilities for version 3.3
2019-04-28 20:39:45.542172 I | embed: ClientTLS: cert = default.etcd/fixtures/client/cert.pem, key = default.etcd/fixtures/client/key.pem, trusted-ca = , client-cert-auth = false, crl-file = 
2019-04-28 20:39:45.542296 I | embed: listening for peers on [::]:2380
{"level":"info","ts":"2019-04-28T20:39:47.158+0100","caller":"raft/raft.go:922","msg":"8e9e05c52164694d is starting a new election at term 35"}
{"level":"info","ts":"2019-04-28T20:39:47.158+0100","caller":"raft/raft.go:725","msg":"8e9e05c52164694d became candidate at term 36"}
{"level":"info","ts":"2019-04-28T20:39:47.158+0100","caller":"raft/raft.go:820","msg":"8e9e05c52164694d received MsgVoteResp from 8e9e05c52164694d at term 36"}
{"level":"info","ts":"2019-04-28T20:39:47.158+0100","caller":"raft/raft.go:777","msg":"8e9e05c52164694d became leader at term 36"}
{"level":"info","ts":"2019-04-28T20:39:47.158+0100","caller":"raft/node.go:330","msg":"raft.node: 8e9e05c52164694d elected leader 8e9e05c52164694d at term 36"}
2019-04-28 20:39:47.159583 I | etcdserver: published {Name:default ClientURLs:[https://127.0.0.1:2379 https://127.0.0.1:4001]} to cluster cdf818194e3a8c32
2019-04-28 20:39:47.159626 I | embed: ready to serve client requests
2019-04-28 20:39:47.159856 I | embed: ready to serve client requests
2019-04-28 20:39:47.192884 I | embed: serving client requests on [::]:4001
2019-04-28 20:39:47.193073 E | etcdmain: accept tcp [::]:4001: use of closed network connection
2019-04-28 20:39:47.199142 I | embed: serving client requests on [::]:2379
WARNING: 2019/04/28 20:39:47 grpc: addrConn.createTransport failed to connect to {0.0.0.0:4001 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: remote error: tls: handshake failure". Reconnecting...
WARNING: 2019/04/28 20:39:47 grpc: addrConn.createTransport failed to connect to {0.0.0.0:2379 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: remote error: tls: handshake failure". Reconnecting...
WARNING: 2019/04/28 20:39:48 grpc: addrConn.createTransport failed to connect to {0.0.0.0:4001 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 0.0.0.0:4001: connect: connection refused". Reconnecting...
WARNING: 2019/04/28 20:39:48 grpc: addrConn.createTransport failed to connect to {0.0.0.0:2379 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 0.0.0.0:2379: connect: connection refused". Reconnecting...
2019-04-28 20:39:48.193328 E | etcdmain: accept tcp [::]:2379: use of closed network connection
2019-04-28 20:39:48.193461 E | etcdmain: http2: TLSConfig.CipherSuites is missing an HTTP/2-required AES_128_GCM_SHA256 cipher.
2019-04-28 20:39:48.193500 E | etcdmain: http2: TLSConfig.CipherSuites is missing an HTTP/2-required AES_128_GCM_SHA256 cipher.
```

### CODE REVIEW

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -287,9 +287,27 @@ func startEtcdOrProxyV2() {
...
-			plog.Fatal(lerr)
+			plog.Error(lerr)
+		}
+		// It may be the case that several listeners have failed at
+		// once. Exiting as soon as the first one fails would
+		// potentially lose valuable diagnostics, so we wait for a
+		// second and log all available errors on errc before exiting.
+		time.Sleep(1 * time.Second)
+
+		for {
```
Comment (xiang90) [inline etcdmain/etcd.go]: we should make the outer select in a for loop, and start to wait for a timeout chan. for timeout, we should assign time.After(x) to a initially nil chan when the first error happens.

Comment (xiang90) [reply etcdmain/etcd.go]: after the timeout chan fires, we can do os.exit(1)

**Comment by rkday** [comment]
Okay - like this?

File: etcdmain/etcd.go [hunk 2]
```diff
@@ -283,15 +283,26 @@ func startEtcdOrProxyV2() {
...
+			if lg != nil {
+				lg.Error("listener failed", zap.Error(lerr))
+			} else {
+				plog.Error(lerr)
+			}
+			// It may be the case that several listeners have failed at
+			// once. Exiting as soon as the first one fails would
+			// potentially lose valuable diagnostics, so we wait for a
+			// second and log all available errors on errc before exiting.
+			timeoutc = time.After(1 * time.Second)
```
Comment (jingyih) [inline etcdmain/etcd.go]: What is the intended behavior?
1. exit the program 1 second after the first error is received? In this case we are collecting errors only within 1 second window.

Or

2. before exiting the program, collect all errors as long as they are not received more than 1 second apart.

Comment (rkday) [reply etcdmain/etcd.go]: Ah - (1) was my intention but this probably does (2), so I probably need a `if timeoutc == nil` or similar.

Comment (jingyih) [reply etcdmain/etcd.go]: Yeah I think the code currently does (2). Checking timeoutc before reassigning sounds good.

File: etcdmain/etcd.go [hunk 3]
```diff
@@ -283,15 +283,26 @@ func startEtcdOrProxyV2() {
...
+		case <-timeoutc:
+			osutil.Exit(1)
+		case <-stopped:
+			osutil.Exit(0)
```
Comment (jingyih) [inline etcdmain/etcd.go]: If exiting here, remove line 308 as it will never be used.

**Pull request closed without merging**